From 38306472594c23d24260014b41bf6259f2f99786 Mon Sep 17 00:00:00 2001
From: Liang Li <liang.li@windriver.com>
Date: Thu, 12 Jul 2012 14:46:19 +0800
Subject: [PATCH] net/core: fix a wrong change introduced by lttng patches

Fix the kernel call trace:
[
.../include/linux/skbuff.h:1314!
invalid opcode: 0000 [#1] PREEMPT SMP
LTT NESTING LEVEL : 0
CPU 2
Modules linked in: hwmon snd_page_alloc coretemp iTCO_wdt ata_generic
container soundcore snd_hda_codec_realtek processor mperf freq_table
acpi_cpufreq snd snd_timer snd_pcm snd_hwdep snd_hda_codec snd_hda_int
el

Pid: 698, comm: dmesg Not tainted
3.4.4-WR5.0+snapshot-20120712_standard #1 Supermicro C2SBC-Q/C2SBC-Q
RIP: 0010:[<ffffffff815fc101>]  [<ffffffff815fc101>]
skb_gso_segment+0x261/0x280
RSP: 0000:ffff88023fd036d0  EFLAGS: 00010287
RAX: ffff8802323a6cce RBX: ffff8802322390e8 RCX: 00000000000000cf
RDX: 000000000000106b RSI: 00000000000000cf RDI: ffff8802322390e8
RBP: ffff88023fd03710 R08: 00000000700049a9 R09: 0000000000000000
R10: 0000000000000002 R11: 0000000000000003 R12: 000000000000000e
R13: 0000000000000008 R14: 00000000700049a9 R15: ffff8802323a6c00
FS:  0000000000000000(0000) GS:ffff88023fd00000(0063)
knlGS:00000000f775d6c0
CS:  0010 DS: 002b ES: 002b CR0: 000000008005003b
CR2: 000000004343d430 CR3: 0000000231ebd000 CR4: 00000000000407e0
DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
DR3: 0000000000000000 DR6: 00000000ffff0ff0 DR7: 0000000000000400
Process dmesg (pid: 698, threadinfo ffff880231cde000, task
ffff880233778000)
Stack:
 ffff8802322386e8 0000000000002000 0000000000000000 ffff8802322390e8
 ffff880232eec000 0000000000000000 ffff880232e86a00 ffff8802323a6c00
 ffff88023fd03790 ffffffff816003eb ffff88023fd03770 ffffffff81634896
Call Trace:
 <IRQ>
 [<ffffffff816003eb>] dev_hard_start_xmit+0x1bb/0x710
 [<ffffffff81634896>] ? ip_finish_output+0x186/0x300
 [<ffffffff8103ac31>] ? raise_softirq_irqoff+0x21/0xa0
 [<ffffffff8161b3fe>] sch_direct_xmit+0xfe/0x1d0
 [<ffffffff81600ae2>] dev_queue_xmit+0x1a2/0x670
 [<ffffffff8163487e>] ip_finish_output+0x16e/0x300
 [<ffffffff81635fe8>] ip_output+0x98/0xa0
 [<ffffffff8162cd6f>] ? ipv4_dst_check+0x2f/0x50
 [<ffffffff816356b9>] ip_local_out+0x29/0x30
 [<ffffffff81635813>] ip_queue_xmit+0x153/0x420
 [<ffffffff8164c5c9>] tcp_transmit_skb+0x3a9/0x890
 [<ffffffff8164d0de>] tcp_write_xmit+0xfe/0xac0
 [<ffffffff815f2857>] ? __kfree_skb+0x47/0xa0
 [<ffffffff8164db06>] __tcp_push_pending_frames+0x26/0x90
 [<ffffffff8164a311>] tcp_rcv_established+0xe1/0x800
 [<ffffffff81650ac5>] tcp_v4_do_rcv+0x125/0x390
 [<ffffffff81653760>] tcp_v4_rcv+0x610/0x910
 [<ffffffff816307fd>] ip_local_deliver_finish+0xed/0x2b0
 [<ffffffff81630ba0>] ip_local_deliver+0x90/0xa0
 [<ffffffff8163049d>] ip_rcv_finish+0x10d/0x380
 [<ffffffff81630e01>] ip_rcv+0x251/0x2f0
 [<ffffffff815fdec6>] __netif_receive_skb+0x4c6/0x610
 [<ffffffff815fe1bd>] netif_receive_skb+0x3d/0x120
 [<ffffffff815fe890>] ? dev_gro_receive+0x1d0/0x2b0
 [<ffffffff815ff030>] napi_skb_finish+0x50/0x70
 [<ffffffff815ff145>] napi_gro_receive+0xf5/0x130
 [<ffffffff81506f19>] e1000_receive_skb+0x59/0x70
 [<ffffffff8150947a>] e1000_clean_rx_irq+0x24a/0x3e0
 [<ffffffff81508c68>] e1000_clean+0x78/0x330
 [<ffffffff815fe503>] net_rx_action+0x103/0x2c0
 [<ffffffff8103a785>] __do_softirq+0xa5/0x200
 [<ffffffff8177c30c>] call_softirq+0x1c/0x30
 [<ffffffff81003935>] do_softirq+0x55/0x90
 [<ffffffff8103abee>] irq_exit+0x8e/0xb0
 [<ffffffff8177c853>] do_IRQ+0x63/0xd0
 [<ffffffff81773ca7>] common_interrupt+0x67/0x67
 <EOI>
 [<ffffffff8177c39b>] ? sysenter_dispatch+0x7/0x1e
Code: d0 83 f8 01 0f 84 7f fe ff ff 31 d2 31 f6 b9 20 00 00 00 48 89
df e8 5f 70 ff ff 85 c0 0f 84 66 fe ff ff 4c 63 e8 e9 04 ff ff ff <0f>
0b 48 83 7a 08 00 0f 84 c9 fe ff ff 41 f6 c6 40 0f 85 bf fe
RIP  [<ffffffff815fc101>] skb_gso_segment+0x261/0x280
 RSP <ffff88023fd036d0>
---[ end trace 999f8e3d4df2aeef ]---
Kernel panic - not syncing: Fatal exception in interrupt
panic occurred, switching back to text console
]

The issue was introduce by LTTNG patches.

Signed-off-by: Liang Li <liang.li@windriver.com>
---
 net/core/dev.c |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index 00c0f00..c6eeba9 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -1968,9 +1968,8 @@ struct sk_buff *skb_gso_segment(struct sk_buff *skb,
 	}
 
 	__this_cpu_inc(softnet_data.processed);
-	skb_reset_network_header(skb);
-	skb_reset_transport_header(skb);
 
+	skb_reset_mac_header(skb);
 	skb->mac_len = skb->network_header - skb->mac_header;
 	__skb_pull(skb, skb->mac_len);
 
-- 
1.7.9.7

